rtt/base/DataObjectLockFree.hpp

```diff
@@ -249,38 +254,49 @@ namespace RTT
          */
         virtual bool Set( param_t push )
         {
-            /**
-             * This method can not be called concurrently (only one
-             * producer). With a minimum of 3 buffers, if the
-             * write_ptr+1 field is not occupied, it will remain so
-             * because the read_ptr is at write_ptr-1 (and can
-             * not increment the counter on write_ptr+1). Hence, no
-             * locking is needed.
-             */
-
             if (!initialized) {
                 log(Error) << "You set a lock-free data object of type " << internal::DataSourceTypeInfo<T>::getType() << " without initializing it with a data sample. "
                            << "This might not be real-time safe." << endlog();
                 data_sample(value_t(), true);
             }
 
-            // writeout in any case
-            PtrType writing = write_ptr;
+            PtrType writing;
+            do {
+                writing = write_ptr;            // copy buffer location
+                oro_atomic_inc(&writing->write_counter); // lock buffer, no more writes
+                // XXX smp_mb
+                if ( writing != write_ptr )     // if write_ptr changed,
+                    oro_atomic_dec(&writing->write_counter); // better to start over.
```

Comment:
Eventually we could simply abort here instead of trying again. Another thread was faster and modified the `write_ptr` already since we made a copy in line 265. At the moment we prefer the sample written by low-priority threads because they will overwrite previous samples after they are unblocked. If `Set()` would abort as soon as it detects attempts to write concurrently, high-priority threads would have a higher chance to "push" their sample.